weld/optimizer/transforms/cse.rs

```diff
@@ -0,0 +1,573 @@
+                    self.generate_bindings_scoped(body, bindings, generated, stack, scopes);
+                    (None, false)
+                }
+                If { ref mut cond, ref mut on_true, ref mut on_false } => {
+                    // Generate bindings for the condition, since we won't recurse down into
+                    // subexpressions.
+                    self.generate_bindings(cond, bindings, generated, stack, scopes);
+
+                    // We want to keep definitions of If/Else statements within the branch target
+                    // to prevent moving expressions out from behind a branch condition.
+                    self.generate_bindings_scoped(on_true, bindings, generated, stack, scopes);
+                    self.generate_bindings_scoped(on_false, bindings, generated, stack, scopes);
+                    (None, false)
+                }
+                Ident(ref mut sym) if !generated.contains(sym) && bindings.contains_key(sym) => {
+                    let mut expr = bindings.get(sym).cloned().unwrap();
+                    generated.insert(sym.clone());
+
+                    // Generate the definition of this identifier if it doesn't exist.
+                    self.generate_bindings(&mut expr, bindings, generated, stack, scopes);
+
+                    trace!("Added binding {} -> {}", &sym, expr.pretty_print());
+                    // We add the generated bindings to the highest scope.
+                    let scope = *scopes.get(&sym).unwrap() as usize;
+                    let binding_list = stack.get_mut(scope).unwrap();
+                    binding_list.push((sym.clone(), expr));
+                    (None, false)
+                }
+                _ => (None, true),
+            }
+        });
+    }
+
+    /// Creates a new scope (e.g., for a function or if statement) and generates bindings.
+    ///
+    /// The new scope represents a new site in the expression tree to insert let statements.
+    ///
+    /// The created bindings may not necessarily be in the newest scope -- bindings for a symbol
+    /// `sym` are generated in the scope defined by `scopes` (i.e., in the scope
+    /// `stack[scopes.get(sym)]`).
+    fn generate_bindings_scoped(&mut self,
+                               expr: &mut Expr,
+                               bindings: &mut HashMap<Symbol, Expr>,
+                               generated: &mut HashSet<Symbol>,
+                               stack: &mut Vec<Vec<(Symbol, Expr)>>,
+                               scopes: &HashMap<Symbol, ScopeDepth>) {
+
+        trace!("Processing scoped expression {}", expr.pretty_print());
+
+        stack.push(vec![]);
+
+        self.generate_bindings(expr, bindings, generated, stack, scopes);
+
+        // Generate the Let expressions for the bindings in this scope.
+        let binding_list = stack.pop().unwrap();
+        let mut prev = expr.take();
+        for (sym, expr) in binding_list.into_iter().rev() {
```

Comment:
Comment on why you're doing this in reverse order. I had to think a while why you did this (because you put all "previous" let expressions in the body of the `let_expr` so they're actually executed after the current let expression?)